Shorthands are still using "initial" for longhands that are set implicitly

darinadler opened the pull request:
#### 03224cfdf4540d4f80ef3c9394815a6fc9ed50b6
<pre>
Shorthands are still using &quot;initial&quot; for longhands that are set implicitly
<a href="https://bugs.webkit.org/show_bug.cgi?id=248913">https://bugs.webkit.org/show_bug.cgi?id=248913</a>
rdar://problem/103092939

Reviewed by Sam Weinig and Oriol Brufau.

This removes all the remaining cases that use &quot;initial&quot; for longhands, moving to initial values
instead, but keeping the initial value placeholder for performance reasons. We now serialize
properties based on checking if longhands have their initial values, and are not relying on the
implicit flag. That&apos;s because in the case of layered shorthands, we can no longer use the implicit
flag; there is nothing that stores those in the layers. We plan to rely on values and not the
implicit flag universally; in another change soon we will finish the job.

I have additional removal and tidying, including renaming createImplicitInitialValue since it
is now a placeholder that represents initial values without having to allocate the memory for
them, rather than a special &quot;implicit&quot; value. But since this is already a large change, holding
those changes for followup.

* LayoutTests/editing/deleting/paste-with-transparent-background-color-expected.txt:
* LayoutTests/editing/deleting/paste-with-transparent-background-color-live-range-expected.txt:
Updated for change where we end up putting in a &quot;background-image: none&quot;. This is a regression,
but not an important one, and improving this to instead emit &quot;background: transparent&quot; or
&quot;background: none&quot; will probably involve improving the editing code rather than the CSS code.

* LayoutTests/editing/pasteboard/copy-paste-inserts-clearing-div-expected.txt:
Updated for progression where we use the background shorthand, not background-color.

* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/css/background-clip-text-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/background-position-serialize-expected.txt: Progression where we serialize
background shorthand instead of background-image and background-position longhands.

* LayoutTests/fast/css/border-shorthand-initialize-longhands-expected.txt:
* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing-expected.txt:
* LayoutTests/fast/css/parsing-text-emphasis-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/remove-shorthand-expected.txt: Expect various shorthands to be handled
correctly, such as adding border-image, handling the background shorthand, and others.

* LayoutTests/fast/css/shorthand-mismatched-list-crash-expected.txt:
Expect test to pass.

* LayoutTests/fast/css/transform-inline-style-expected.txt: Don&apos;t expect initial values &quot;ease&quot; and
&quot;0s&quot; to be serialized.
* LayoutTests/fast/css/transform-inline-style-remove-expected.txt: Ditto, also initial values
&quot;none&quot; and &quot;running&quot;.

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt:
* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/style-src/inline-style-allowed-while-cloning-objects.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/border-shorthand-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Expect more tests to pass.

* Source/WebCore/css/CSSProperties.json: Added the border-image longhands to the border shorthand,
because the border shorthand sets all these properties to initial values. Also sort the offset
shorthand longhands in the order from the specification.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::textExpandingInitialValuePlaceholder): Added.
(WebCore::valueIDExpandingInitialValuePlaceholder): Added.
(WebCore::StyleProperties::getPropertyValue const): Added a call to textExpandingInitialValuePlaceholder
so we expand the shorthand rather than serializing &quot;initial&quot;. Removed the borderSpacingValue function
and use the get2Values function instead for border-spacing. If border-image shorthand values don&apos;t all
have their initial values, then do not serialize the border shorthand. Use getPropertyValue instead of
getPropertyCSSValue and cssText so we get the additional initial value placeholder logic.
(WebCore::StyleProperties::propertyAsColor const): Use textExpandingInitialValuePlaceholder.
(WebCore::StyleProperties::propertyAsValueID const): Use valueIDExpandingInitialValuePlaceholder.
(WebCore::StyleProperties::borderSpacingValue const): Deleted.
(WebCore::StyleProperties::fontValue const):  Use textExpandingInitialValuePlaceholder.
(WebCore::StyleProperties::fontVariantValue const): Check isImplicitInitialValue since that&apos;s a second
way the value normal could be specified.
(WebCore::StyleProperties::get2Values const): Rewrote more simply, removing the isInitialValue
checks since commonShorthandChecks takes care of that now. Also use makeString instead of
StringBuilder since we can and it&apos;s more efficient.
(WebCore::StyleProperties::get4Values const): Rewrote more simply, removing the isInherited
and isInitialValue checks since commonShorthandChecks takes care of that now. Also use makeString
instead of StringBuilder since we can and it&apos;s more efficient.
(WebCore::LayerValues::LayerValues): Added.
(WebCore::LayerValues::set): Added.
(WebCore::LayerValues::skip): Added.
(WebCore::LayerValues::valueID const): Added.
(WebCore::LayerValues::valueIDIncludingCustomIdent const): Added.
(WebCore::LayerValues::equalValueIDs const): Added.
(WebCore::LayerValues::isValueID const): Added.
(WebCore::LayerValues::isPair const): Added.
(WebCore::LayerValues::serialize const): Added.
(WebCore::StyleProperties::getLayeredShorthandValue const): Rewrote this to use initial values
instead of relying on implicit flags to know what to not serialize. Added various rules that were
not present in the old version that are now needed for round tripping and omitting any value that
can be without changing the meaning. The LayerValues class above is part of this rewrite.
(WebCore::StyleProperties::getShorthandValue const): Use getPropertyValue instead of
getPropertyCSSValue and cssText so we get the additional initial value placeholder logic.
(WebCore::StyleProperties::getCommonValue const):  Ditto.
(WebCore::StyleProperties::borderImagePropertyValue const): Serialize as &quot;none&quot; when everything
has the implicit flag set.
(WebCore::StyleProperties::borderPropertyValue const): Rewrote for simplicitly. Now checks for
initial values instead of depending on the implicit flag, using string comparisons since that&apos;s
the easiest way to build on getCommonValue. Also use makeString instead of StringBuilder.
(WebCore::StyleProperties::pageBreakValue const): Use nameString for slightly better efficiency.
(WebCore::StyleProperties::webkitColumnBreakValue const): Ditto.
(WebCore::StyleProperties::hasAllInitialValues const): Added.
(WebCore::canUseShorthandForLonghand): Allow use of the CSSPropertyBackground shorthand,
which did not work well before getLayeredShorthandValue omitted initial values.
(WebCore::StyleProperties::asTextInternal const): Removed now-incorrect handling of
isImplicitInitialValue and use textExpandingInitialValuePlaceholder.

* Source/WebCore/css/StyleProperties.h: Added hasAllInitialValues. Removed borderSpacingValue.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::addProperty): Merged addProperty and addPropertyWithImplicitDefault
into a single function, which can take nullptr and replace it with the initial value placeholder.
(WebCore::CSSPropertyParser::addExpandedProperty): Renamed from addExpandedPropertyForValue,
changed to take nullptr like addProperty now does, and added an implicit flag.
(WebCore::CSSPropertyParser::parseValueStart): Updated for addProperty and addExpandedProperty
changes.
(WebCore::CSSPropertyParser::consumeCSSWideKeyword): Ditto.
(WebCore::CSSPropertyParser::parseCounterStyleDescriptor): Ditto.
(WebCore::CSSPropertyParser::parseFontFaceDescriptor): Ditto.
(WebCore::CSSPropertyParser::parsePropertyDescriptor): Ditto.
(WebCore::CSSPropertyParser::parseFontPaletteValuesDescriptor): Ditto.
(WebCore::CSSPropertyParser::consumeFontVariantShorthand): Ditto.
(WebCore::CSSPropertyParser::consumeColumns): Ditto.
(WebCore::initialValueForLonghand): Added.
(WebCore::initialCSSValueForAnimationLonghand): Added.
(WebCore::isNumber): Added.
(WebCore::pairValue): Added.
(WebCore::quadValue): Added.
(WebCore::isValueIDPair): Added.
(WebCore::isNumericQuad): Added.
(WebCore::isInitialValueForLonghand): Added.
(WebCore::initialValueTextForLonghand): Added.
(WebCore::initialValueIDForLonghand): Added.
(WebCore::CSSPropertyParser::consumeShorthandGreedily): Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeBorderShorthand): Renamed from consumeBorder. Do the calls
to addExpandedProperty here instead of in the caller. Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeBorderImage): Udpated for addProperty changes. This removed
the need to create value objects since we now use the initial value placeholder.
(WebCore::CSSPropertyParser::consumeAnimationShorthand): Use initialCSSValueForAnimationLonghand
instead of ComputedStyleExtractor::addValueForAnimationPropertyToList. Updated for addProperty
changes.
(WebCore::CSSPropertyParser::consumeBackgroundShorthand): Removed local variable &quot;implicit&quot;
that was always false. Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeOverscrollBehaviorShorthand): Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeContainerShorthand): Ditto.
(WebCore::CSSPropertyParser::consumeContainIntrinsicSizeShorthand): Ditto.
(WebCore::CSSPropertyParser::consumeTransformOrigin): Ditto.
(WebCore::CSSPropertyParser::consumeOffset): Ditto.
(WebCore::CSSPropertyParser::consumeListStyleShorthand): Ditto.
(WebCore::CSSPropertyParser::parseShorthand): Use consumeBorderShorthand. For CSSPropertyBorder
itself, also set all the border-image longhands to their initial values.

* Source/WebCore/css/parser/CSSPropertyParser.h: Updated the addProperty and addExpandedProperty
functions. Replaced consumeBorder with consumeBorderShorthand. Added isInitialValueForLonghand,
initialValueTextForLonghand, and intialValueIDForLonghand, exported by the parser but used by the
serialization code.

Canonical link: <a href="https://commits.webkit.org/258767@main">https://commits.webkit.org/258767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51fd05c62b9225e616bee80ddb58899a340e3ff2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102890 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112141 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172361 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106851 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13025 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2914 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95134 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108664 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/24732 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5457 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/26146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2598 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11620 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45638 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6027 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7354 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on LayoutTests/editing/deleting/paste-with-transparent-background-color-expected.txt:
```diff
@@ -13,7 +13,7 @@ After cut and paste:
 | <span>
 |   class="test"
-|   style="background-color: transparent;"
+|   style="background-image: none; background-position: 0%; background-size: auto; background-repeat: repeat; background-attachment: scroll; background-origin: padding-box; background-clip: border-box; background-color: transparent;"
```
`CSSPropertyBackground` is no longer returning false for `canUseShorthandForLonghand`, so why isn't the shorthand used?

darinadler replied:
I accidentally rebased this test while I had a change in the tree. The expected value here is wrong. Easy to fix.

Loirooriol reviewed (commented):
The patch is quite big, and there might be risk for another revert, so it may be good to try to split it into smaller parts.

Like some changes are just formatting, or things like including `border-image-*` in the list of `border` longhands could be done on their own.

Loirooriol commented on LayoutTests/editing/deleting/paste-with-transparent-background-color-expected.txt:
```diff
@@ -13,7 +13,7 @@ After cut and paste:
 | <span>
 |   class="test"
-|   style="background-color: transparent;"
+|   style="background-image: none; background-color: transparent;"
```
Shouldn't this stay as `background: transparent`?

darinadler replied:
Yes, that’s what we’d want the editing code to do. I don’t know how to quickly fix this right now, but I do at least know after a lot of research that the change will be to the editing code, not the CSS or style system code.

I think the core of the problem is that the editing code, for brevity, is trying to leave out various properties since they are the same as the surrounding context. But I think that for true brevity and best behavior we would want it do that only after taking shorthands into account. Right now, I think it does everything in terms of the longhands. But I honestly wasn’t able to quickly figure out _exactly_ why we end up with this, instead of just `background: transparent` or even just `background-color: transparent`.

Loirooriol commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -892,12 +885,314 @@ bool CSSPropertyParser::consumeColumns(bool important)
+    case CSSPropertyTransitionProperty:
+        return CSSValueAll;
+    default:
+        return { };
```
Maybe add `ASSERT_NOT_REACHED()` here instead of in the consumers?

darinadler replied:
Excellent suggestion. I didn’t notice this potential had revealed itself while refactoring, but I can make it `RELEASE_ASSERT_NOT_REACHED` and remove `monostate` from the variant entirely, simplifying every call site even more.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -580,180 +575,272 @@ String StyleProperties::fontSynthesisValue() const
+                    if (i)
+                        skipSerializing = true;
+                } else {
+                    // Color is only in the last layer.
```
I think this would be slightly cleared placing this comment closer to the condition with `CSSPropertyBackgroundColor`, i.e.

```cpp
if (shorthand.properties()[j] == CSSPropertyBackgroundColor) {
    // Color is only in the last layer.
    ...
} else {
````

darinadler replied:
Sure, I’ll change the `if` statement back to help that read better. Not sure why I had reversed it.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -580,180 +575,272 @@ String StyleProperties::fontSynthesisValue() const
+                case CSSValueEaseIn:
+                case CSSValueEaseInOut:
+                case CSSValueEaseOut:
+                case CSSValueLinear:
```
Missing `CSSValueStepStart` and `CSSValueStepEnd`?

darinadler replied:
Maybe I looked at an older version of the specification; somehow I missed those. Will add them. I guess what we really need is tests covering those cases.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -580,180 +575,272 @@ String StyleProperties::fontSynthesisValue() const
+        ASSERT(indexA < m_shorthand.length());
+        ASSERT(indexB < m_shorthand.length());
+        auto valueA = valueIDExpandingInitialValuePlaceholder(m_shorthand.properties()[indexA], m_values[indexA].get());
+        auto valueB = valueIDExpandingInitialValuePlaceholder(m_shorthand.properties()[indexB], m_values[indexB].get());
```
Could be just

```cpp
auto valueA = valueID(indexA);
auto valueB = valueID(indexB);
```

darinadler replied:
Great idea. I’ll change it to that.

Loirooriol reviewed (approved):
LGTM with nits

Loirooriol commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -892,12 +885,306 @@ bool CSSPropertyParser::consumeColumns(bool important)
+        }
+        ASSERT_NOT_REACHED();
+        return ""_s;
+    }, [](std::monostate) {
+        ASSERT_NOT_REACHED();
+        return ""_s;
+    });
```
I think you can remove this

darinadler replied:
Yes, I’ll clean up after landing. Surprised it even compiles!

darinadler committed:
Shorthands are still using "initial" for longhands that are set implicitly
https://bugs.webkit.org/show_bug.cgi?id=248913
rdar://problem/103092939

Reviewed by Sam Weinig and Oriol Brufau.

This removes all the remaining cases that use "initial" for longhands, moving to initial values
instead, but keeping the initial value placeholder for performance reasons. We now serialize
properties based on checking if longhands have their initial values, and are not relying on the
implicit flag. That's because in the case of layered shorthands, we can no longer use the implicit
flag; there is nothing that stores those in the layers. We plan to rely on values and not the
implicit flag universally; in another change soon we will finish the job.

I have additional removal and tidying, including renaming createImplicitInitialValue since it
is now a placeholder that represents initial values without having to allocate the memory for
them, rather than a special "implicit" value. But since this is already a large change, holding
those changes for followup.

* LayoutTests/editing/deleting/paste-with-transparent-background-color-expected.txt:
* LayoutTests/editing/deleting/paste-with-transparent-background-color-live-range-expected.txt:
Updated for change where we end up putting in a "background-image: none". This is a regression,
but not an important one, and improving this to instead emit "background: transparent" or
"background: none" will probably involve improving the editing code rather than the CSS code.

* LayoutTests/editing/pasteboard/copy-paste-inserts-clearing-div-expected.txt:
Updated for progression where we use the background shorthand, not background-color.

* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/css/background-clip-text-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/background-position-serialize-expected.txt: Progression where we serialize
background shorthand instead of background-image and background-position longhands.

* LayoutTests/fast/css/border-shorthand-initialize-longhands-expected.txt:
* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing-expected.txt:
* LayoutTests/fast/css/parsing-text-emphasis-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/remove-shorthand-expected.txt: Expect various shorthands to be handled
correctly, such as adding border-image, handling the background shorthand, and others.

* LayoutTests/fast/css/shorthand-mismatched-list-crash-expected.txt:
Expect test to pass.

* LayoutTests/fast/css/transform-inline-style-expected.txt: Don't expect initial values "ease" and
"0s" to be serialized.
* LayoutTests/fast/css/transform-inline-style-remove-expected.txt: Ditto, also initial values
"none" and "running".

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt:
* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/style-src/inline-style-allowed-while-cloning-objects.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/border-shorthand-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Expect more tests to pass.

* Source/WebCore/css/CSSProperties.json: Added the border-image longhands to the border shorthand,
because the border shorthand sets all these properties to initial values. Also sort the offset
shorthand longhands in the order from the specification.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::textExpandingInitialValuePlaceholder): Added.
(WebCore::valueIDExpandingInitialValuePlaceholder): Added.
(WebCore::StyleProperties::getPropertyValue const): Added a call to textExpandingInitialValuePlaceholder
so we expand the shorthand rather than serializing "initial". Removed the borderSpacingValue function
and use the get2Values function instead for border-spacing. If border-image shorthand values don't all
have their initial values, then do not serialize the border shorthand. Use getPropertyValue instead of
getPropertyCSSValue and cssText so we get the additional initial value placeholder logic.
(WebCore::StyleProperties::propertyAsColor const): Use textExpandingInitialValuePlaceholder.
(WebCore::StyleProperties::propertyAsValueID const): Use valueIDExpandingInitialValuePlaceholder.
(WebCore::StyleProperties::borderSpacingValue const): Deleted.
(WebCore::StyleProperties::fontValue const):  Use textExpandingInitialValuePlaceholder.
(WebCore::StyleProperties::fontVariantValue const): Check isImplicitInitialValue since that's a second
way the value normal could be specified.
(WebCore::StyleProperties::get2Values const): Rewrote more simply, removing the isInitialValue
checks since commonShorthandChecks takes care of that now. Also use makeString instead of
StringBuilder since we can and it's more efficient.
(WebCore::StyleProperties::get4Values const): Rewrote more simply, removing the isInherited
and isInitialValue checks since commonShorthandChecks takes care of that now. Also use makeString
instead of StringBuilder since we can and it's more efficient.
(WebCore::LayerValues::LayerValues): Added.
(WebCore::LayerValues::set): Added.
(WebCore::LayerValues::skip): Added.
(WebCore::LayerValues::valueID const): Added.
(WebCore::LayerValues::valueIDIncludingCustomIdent const): Added.
(WebCore::LayerValues::equalValueIDs const): Added.
(WebCore::LayerValues::isValueID const): Added.
(WebCore::LayerValues::isPair const): Added.
(WebCore::LayerValues::serialize const): Added.
(WebCore::StyleProperties::getLayeredShorthandValue const): Rewrote this to use initial values
instead of relying on implicit flags to know what to not serialize. Added various rules that were
not present in the old version that are now needed for round tripping and omitting any value that
can be without changing the meaning. The LayerValues class above is part of this rewrite.
(WebCore::StyleProperties::getShorthandValue const): Use getPropertyValue instead of
getPropertyCSSValue and cssText so we get the additional initial value placeholder logic.
(WebCore::StyleProperties::getCommonValue const):  Ditto.
(WebCore::StyleProperties::borderImagePropertyValue const): Serialize as "none" when everything
has the implicit flag set.
(WebCore::StyleProperties::borderPropertyValue const): Rewrote for simplicitly. Now checks for
initial values instead of depending on the implicit flag, using string comparisons since that's
the easiest way to build on getCommonValue. Also use makeString instead of StringBuilder.
(WebCore::StyleProperties::pageBreakValue const): Use nameString for slightly better efficiency.
(WebCore::StyleProperties::webkitColumnBreakValue const): Ditto.
(WebCore::StyleProperties::hasAllInitialValues const): Added.
(WebCore::canUseShorthandForLonghand): Allow use of the CSSPropertyBackground shorthand,
which did not work well before getLayeredShorthandValue omitted initial values.
(WebCore::StyleProperties::asTextInternal const): Removed now-incorrect handling of
isImplicitInitialValue and use textExpandingInitialValuePlaceholder.

* Source/WebCore/css/StyleProperties.h: Added hasAllInitialValues. Removed borderSpacingValue.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::addProperty): Merged addProperty and addPropertyWithImplicitDefault
into a single function, which can take nullptr and replace it with the initial value placeholder.
(WebCore::CSSPropertyParser::addExpandedProperty): Renamed from addExpandedPropertyForValue,
changed to take nullptr like addProperty now does, and added an implicit flag.
(WebCore::CSSPropertyParser::parseValueStart): Updated for addProperty and addExpandedProperty
changes.
(WebCore::CSSPropertyParser::consumeCSSWideKeyword): Ditto.
(WebCore::CSSPropertyParser::parseCounterStyleDescriptor): Ditto.
(WebCore::CSSPropertyParser::parseFontFaceDescriptor): Ditto.
(WebCore::CSSPropertyParser::parsePropertyDescriptor): Ditto.
(WebCore::CSSPropertyParser::parseFontPaletteValuesDescriptor): Ditto.
(WebCore::CSSPropertyParser::consumeFontVariantShorthand): Ditto.
(WebCore::CSSPropertyParser::consumeColumns): Ditto.
(WebCore::initialValueForLonghand): Added.
(WebCore::initialCSSValueForAnimationLonghand): Added.
(WebCore::isNumber): Added.
(WebCore::pairValue): Added.
(WebCore::quadValue): Added.
(WebCore::isValueIDPair): Added.
(WebCore::isNumericQuad): Added.
(WebCore::isInitialValueForLonghand): Added.
(WebCore::initialValueTextForLonghand): Added.
(WebCore::initialValueIDForLonghand): Added.
(WebCore::CSSPropertyParser::consumeShorthandGreedily): Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeBorderShorthand): Renamed from consumeBorder. Do the calls
to addExpandedProperty here instead of in the caller. Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeBorderImage): Udpated for addProperty changes. This removed
the need to create value objects since we now use the initial value placeholder.
(WebCore::CSSPropertyParser::consumeAnimationShorthand): Use initialCSSValueForAnimationLonghand
instead of ComputedStyleExtractor::addValueForAnimationPropertyToList. Updated for addProperty
changes.
(WebCore::CSSPropertyParser::consumeBackgroundShorthand): Removed local variable "implicit"
that was always false. Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeOverscrollBehaviorShorthand): Updated for addProperty changes.
(WebCore::CSSPropertyParser::consumeContainerShorthand): Ditto.
(WebCore::CSSPropertyParser::consumeContainIntrinsicSizeShorthand): Ditto.
(WebCore::CSSPropertyParser::consumeTransformOrigin): Ditto.
(WebCore::CSSPropertyParser::consumeOffset): Ditto.
(WebCore::CSSPropertyParser::consumeListStyleShorthand): Ditto.
(WebCore::CSSPropertyParser::parseShorthand): Use consumeBorderShorthand. For CSSPropertyBorder
itself, also set all the border-image longhands to their initial values.

* Source/WebCore/css/parser/CSSPropertyParser.h: Updated the addProperty and addExpandedProperty
functions. Replaced consumeBorder with consumeBorderShorthand. Added isInitialValueForLonghand,
initialValueTextForLonghand, and intialValueIDForLonghand, exported by the parser but used by the
serialization code.

Canonical link: https://commits.webkit.org/258767@main
